sophia.man
.\"
.\" DI $Id: sophia.tr,v 1.10 2014/03/21 09:48:19 pj Exp pj $
.\" DA Pierre‐Jean Fichet
.\" DS Utroff sophia manual
.\" DT Utroff sophia manual
.\" DK utroff sophia philia troff nroff heirloom tmac xml sqlite
.
.
.
.TH SOPHIA 1 ´2018‐04‐10´
.
.
.
.SH NAME
.PP
\fISophia\fR – Maintain a database of texts.
.
.
.
.SH SYNOPSIS
.PP
\fIsophia\fR [
\fBinsert\fR |
\fBsearch\fR |
\fBedit\fR   |
\fBupdate\fR |
\fBrefer\fR  |
\fBtroff\fR  |
\fBprint\fR ]
[
\fB‐b\fR \fIbookid\fR
\fB‐ba\fR \fIdata\fR
\fB‐bb\fR \fIdata\fR
\fB‐bc\fR \fIdata\fR
\fB‐bd\fR \fIdata\fR
\fB‐be\fR \fIdata\fR
\fB‐bg\fR \fIdata\fR
\fB‐bh\fR \fIdata\fR
\fB‐bi\fR \fIdata\fR
\fB‐bj\fR \fIdata\fR
\fB‐bm\fR \fIdata\fR
\fB‐bn\fR \fIdata\fR
\fB‐bo\fR \fIdata\fR
\fB‐bq\fR \fIdata\fR
\fB‐bs\fR \fIdata\fR
\fB‐bt\fR \fIdata\fR
\fB‐bx\fR \fIdata\fR
\fB‐by\fR \fIdata\fR
\fB‐bz\fR \fIdata\fR
\fB‐h\fR \fIheadid\fR
\fB‐hb\fR \fIdata\fR
\fB‐hh\fR \fIdata\fR
\fB‐hp\fR \fIdata\fR
\fB‐t\fR \fItextid\fR
\fB‐th\fR \fIdata\fR
\fB‐ts\fR \fIdata\fR
\fB‐tt\fR \fIdata\fR
\fB‐tp\fR \fIdata\fR
]
.
.
.
.SH DESCRIPTION
.PP
\fISophia\fR maintains a database of texts: It allows to
easyly insert, modify, search and extract information to and
from the database. Made to write a philosophy thesis,
\fIsophia\fR has two aims: comparing texts, and producing
precise references:
.TP
\&Comparing text:
To write a philosophy thesis, one have to compare hundreds
of texts. This can be done by turning pages of books, but
this is not conveniant when writing with a computer. The
author of \fIsophia\fR found easier to write all the texts
he had to work with in a database, and then find and
compare them from the computer, just near his text editor.
.TP
\&Producing precise references:
While inserting references in a document, the usage is to
mention the heading in which the reference appears. This is
not easyly done with usual bibliography software such as
\fIrefer\fR or \fIbibtex\fR: The heading has to be
duplicated for each reference, which is fastidious and leads
to typo errors and glitches. With \fISophia\fR, each heading
is written once, and is easyly recovered and
automatically inserted in the text document.
.PP
\fISophia\fR is just an interface to the database of texts.
To use it, the format of the database must be understood.
The first argument of \fISophia\fR is a \fBcommand\fR which tell
to sophia what to do: insert, update, search... The
following arguments of sophia correspond to the \fBfields\fR of
the database, and must be followed by the \fIdatas\fR that has to
be inserted, updated, searched...
.
.
.
.SH COMMANDS
.TP
\&\fBedit\fR [\fB‐b\fR|\fB‐h\fR|\fB‐t\fR \fIid\fR] [\fB‐field\fR \fIdata\fR [\fB‐field\fR \fIdata\fR...]]
Call a text editor to edit a row of the database. If [\fB‐b\fR|\fB‐h\fR|\fB‐t\fR \fIid\fR]
option is given, the content of the fields are extracted from
the database. See \fBTEXT FORMAT\fR to know how to insert
information in the text editor.
.TP
\&\fBhead\fR \fB‐hb\fR \fIid\fR \fB‐tp\fR \fIpage\fR
Search the heading of the page \fIpage\fR in the book of
unique id \fIid\fR. Print the unique id of the row of the
table \fBheads\fR where \fBhb\fR is \fIid\fR and where
\fBhp\fR is less than \fBtp\fR and is more than all the
other \fBhp\fR fields.
.TP
\&\fBinsert\fR \fB‐field\fR \fIdata\fR [\fB‐field\fR \fIdata\fR...]
Create a new row in the database by inserting \fIdata\fR in
field \fBfield\fR.
.TP
\&\fBprint\fR \fB‐field\fR \fIdata\fR [\fB‐field\fR \fIdata\fR...]
Print the whole rows of the database where \fBfield\fR contain
\fIdata\fR.
.TP
\&\fBrefer\fR
Output the \fBbooks\fR table of the database in the
\fIrefer\fR bibliography format.
.TP
\&\fBsearch\fR \fB‐field\fR \fIdata\fR [\fB‐field\fR \fIdata\fR...]
Search row where \fBfield\fR contain \fIdata\fR in the
database, and print a summary of their content.
.TP
\&\fBsophia\fR \fB‐field\fR \fIdata\fR [\fB‐field\fR \fIdata\fR...]
Output row where \fBfield\fR contain \fIdata\fR in the
format of the \fIsophia\fR command line. Usefull for
scripting.
.TP
\&\fBtroff\fR \fB‐field\fR \fIdata\fR [\fB‐field\fR \fIdata\fR...]
Print the content of row where \fBfield\fR contain
\fIdata\fR in a format that troff can format.
.TP
\&\fBupdate\fR \fB‐b\fR|\fB‐h\fR|\fB‐t\fR \fIid\fR \fB−field\fR \fIdata\fR
Update the database, by replacing the content of \fBfield\fR
by \fIdata\fR. The \fBunique identifier\fR of the row
which is modified \fBmust\fR be mentioned, by the options
\fB‐b\fR for table \fBbooks\fR, \fB‐h\fR for table
\fBheads\fR, \fB‐t\fR for table \fBtexts\fR.
.
.
.
.SH FIELDS
.PP
The sophia database contain three tables: \fIbooks\fR,
\fIheads\fR and \fItexts\fR. Each text in the table
\fItexts\fR is linked to a heading in the table
\fIheads\fR, and each heading in the table \fIheads\fR is
linked to a book in the table \fIbooks\fR.
.
.
.
.SS Table books fields
.RS 3
.TP
\&\fBb\fR
Unique identifier of the row, automatically generated by
\fIsophia\fR.
.TP
\&\fBba\fR
(field \fB%A\fR of refer bibliography)
Author name and lastname.
.TP
\&\fBbb\fR
(field \fB%B\fR of refer bibliography)
Book title, if the reference is an article in a book.
.TP
\&\fBbc\fR
(field \fB%C\fR of refer bibliography)
City.
.TP
\&\fBbd\fR
(field \fB%D\fR of refer bibliography)
Date
.TP
\&\fBbe\fR
(field \fB%E\fR of refer bibliography)
Editor of the book.
.TP
\&\fBbg\fR
(field \fB%G\fR of refer bibliography)
.TP
\&\fBbh\fR
(field \fB%H\fR of refer bibliography)
How is published the book (first edition, french
edition...).
.TP
\&\fBbi\fR
(field \fB%I\fR of refer bibliography)
Issuer.
.TP
\&\fBbj\fR
(field \fB%J\fR of refer bibliography)
Title if this is a journal.
.TP
\&\fBbm\fR
(field \fB%M\fR of refer bibliography)
Collection.
.TP
\&\fBbn\fR
(field \fB%N\fR of refer bibliography)
Number of the journal
.TP
\&\fBbo\fR
(field \fB%O\fR of refer bibliography)
Other informations.
.TP
\&\fBbq\fR
(field \fB%Q\fR of refer bibliography)
Institution author name.
.TP
\&\fBbs\fR
(field \fB%S\fR of refer bibliography)
Serie name, if the book is part of a serie.
.TP
\&\fBbt\fR
(field \fB%T\fR of refer bibliography)
Title.
.TP
\&\fBbx\fR
(field \fB%x\fR of refer bibliography)
url.
.TP
\&\fBby\fR
(field \fB%y\fR of refer bibliography)
Last update.
.TP
\&\fBby\fR
(field \fB%z\fR of refer bibliography)
Last seen.
.
.
.
.SS Table heads fields
.RS 3
.TP
\&\fBh\fR
Unique identifier of the row, automatically generated by
\fIsophia\fR.
.TP
\&\fBhb\fR
Book unique identifier (link to the fiel \fBb\fR of the
\fIbooks\fR table).
.TP
\&\fBhh\fR
(field \fB%h\fR of refer bibliography)
Heading title. You should enter each heading of the table of
contents, since this is needed for your reader to find the
exact reference in different issues.
.TP
\&\fBhp\fR
Heading page. \fISophia\fR \fBhead\fR option need this field
to find the heading unique id of a page of the book.
.
.
.
.SS Table texts fields
.RS 3
.TP
\&\fBt\fR
Unique identifier of the row, automatically generated by
\fIsophia\fR.
.TP
\&\fBth\fR
Heading unique identifier (link to the file \fBh\fR of the
\fIheads\fR table).
.TP
\&\fBts\fR
Subtitle. Title that you choose for the text.
.TP
\&\fBtt\fR
Text content. The text must begin by troff paragraph macro (such
as \fB.PP\fR, \fBPQ\fR, \fBPI\fR...), or, at least, by
\fB.br\fR.
.TP
\&\fBtp\fR
(field \fB%p\fR of refer bibliography)
Page of the text. A row of pages can be noted as p‐q.
.
.
.
.SH ENVIRONMENT
.PP
\fBSOPHIA\fR the database.
.
.
.
.SH LICENSE
.PP
\fISophia\fR and this manual page are distributed under a
two clause BSD license.
.
.
.
.SH AUTHOR
.PP
Pierre‐Jean Fichet.


sophia.py
#! /usr/bin/env python3
# -*- coding: UTF-8 -*-

"""
module sophia
manage a database of texts

import sophia_class
so=sophia_class.sophia(database)
	# default database is sophia.db
so.insert(dict)
	# return a list containing one dict of datas inserted
so.edit(dict)
	# dict must contain "id:value"
	# return a list containing one dict of datas as they are after edition
so.search(dict)
	# return a list of dicts
so.print(dict)
	# return a list of dict of the content of the whole table
	# or return a list of dict of the selected row

dict has the form {key:value}

One of the keys may define the table:
{ table:books } or { table:texts }
But usually, sophia will deduce it from the keys.

In table books, keys are:
{ id: auto increment id,
A:author, B:book, C:city, D:date, E:editor, G:ISBN/ISSN,
H:publication, I:issuer, J:journal, K:keyword, M:collection
N:journal number, O:other, P:pages of book
Q:author if author is not a person, R:report number
S:serie title, T:title, U:co-author, V:volume
w:electronic access type, x:url, y:update, z:last access } 

In table texts, keys are:
{ id:auto increment id, bid:id of book,
ts:title of this text, tg:some tags to easily search,
th:chapter heading of this text, tp:page of this text,
tt:the text }

Some fields are used by refer, but not by sophia:
F:number of reference, L:label, X: annotation,
o:first occurence of citation in isobib

"""


# depends on: python, sqlite3
import os.path
import sqlite3


# optionnaly depends on psyco
try:
	import psyco
	psyco.full()
except ImportError:
	pass

class TableError(Exception):
	"""	Exception concerning table:
		No table with the column x
		Cannot insert in table alls
		Cannot modify table alls
		You must specify an id to edit
	"""

	def __init__(self, text):
		self.text = text

	def __str__(self):
		return self.text


class sophia:
	""" Simple interface to deal with a database of texts.
		User can use functions: insert(), edit(), search(), print()
		But should'nt use: sqlite(), deftable(), create(), makedict()
	"""

	def __init__(self, db="sophia.db"):
		""" init variables """

		if db != "sophia.db":
			self.db = db
		else:
			if not os.path.exists(db):
				try:
					self.db = os.environ["SOPHIA_DB"]
				except KeyError:
					self.db = "sophia.db"
			else:
				self.db = db

		self.tables = {}
		self.tables["texts"] = ("id", "hid", "ts", "tg", "tp", "tt")
		self.tables["heads"] = ("id", "bid", "hh", "hp")
		self.tables["books"] = ("id", "A", "A2", "A3", "B", "C", "D", "E",
		"E2", "E3", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P",
		"Q", "R", "S", "T", "U", "V", "w", "x", "y", "z")
		# alls is texts + books with only one id, so remove first item of books
		self.tables["alls"] = self.tables["texts"] + self.tables["heads"][1:] \
		+ self.tables["books"][1:]

		if not os.path.exists(self.db):
			self.create()


	def sqlite(self, query, values=None):
		"""Shortcut to execute sqlite query"""
		con = sqlite3.connect(self.db)
		cur=con.cursor()
		# use the following to get a case sensitive database:
		#cur.execute("PRAGMA case_sensitive_like = 1;")
		#con.commit()
		if values == None:
			cur.execute(query)
		else:
			cur.execute(query, values)
		self.lastid=cur.lastrowid
		self.result = cur.fetchall()
		con.commit()
		con.close()


	def create(self):
		"""Create database"""
		view = "CREATE VIEW alls AS SELECT\n texts.id, "
		for table in ("texts", "heads", "books"):
			query = "CREATE TABLE " + table
			query = query + "\n (id INTEGER PRIMARY KEY, "
			# id is yet defined in query, so remove it with [1:]
			for column in self.tables[table][1:]:
				if column == "hid" or column == "bid" or column == "hp" :
					coltype = " INTEGER, "
				else:
					coltype = " TEXT, "
				query = query + column + coltype
				view = view + table + "." + column + ", "
			query = query[:-2] + ");"
			self.sqlite(query)
		view = view[:-2] + """
FROM texts, heads, books
WHERE texts.hid = heads.id AND heads.bid = books.id
ORDER BY A, E, T, texts.id;"""
		self.sqlite(view)

		print ("Database", self.db, "created.")


	def deftable(self, indict):
		"""Define table using indict"""
		self.table=""
		# table might be defined in indict:
		try:
			self.table = indict["table"]
			del indict["table"]
		except KeyError:
			pass
		# always verify that fields are valids:
		for key in indict:
			if key != "id":
				if self.table != "alls" and self.table != "books" \
				and self.table != "heads" and key in self.tables["texts"] :
					self.table = "texts"
				elif self.table != "alls" and self.table != "texts" \
				and self.table != "heads" and key in self.tables["books"] :
					self.table = "books"
				elif self.table != "alls" and self.table != "texts" \
				and self.table != "books" and key in self.tables["heads"] :
					self.table = "heads"
				elif key in self.tables["alls"]:
					self.table = "alls"
				else:
					raise TableError("No table with the column", key)
		if self.table == "":
				self.table = "alls"

		
	def makedict(self):
		"""Make a list of dict to easyly print results"""
		outlist=[]
		for row in self.result:
			outdict= {}
			outdict["table"] = self.table
			i = 0
			for key in self.tables[self.table]:
				if row[i] != None:
					outdict[key] = row[i]
				i = i + 1
			outlist.append(outdict)
		return outlist


	def insert(self, indict):
		"""Insert into table"""
		self.deftable(indict)
		if self.table == "alls":
			raise TableError("Cannot insert in table alls")
		query = "INSERT INTO " + self.table + " (" 
		query_val = ""
		values = []
		for key in indict:
			query = query + key + ", "
			query_val = query_val + "?,"
			values.append(indict[key])
		query = query[:-2] + ") VALUES (" + query_val[:-1] + ");"
		self.sqlite(query, values)

		#indict['id'] = self.lastid
		#print(indict)
		return self.print({'table':self.table, 'id':self.lastid})

	def edit(self, indict):
		"""Update table"""
		self.deftable(indict)
		if self.table == "alls":
			raise TableError("Cannot modify table alls")
		try:
			id = indict["id"]
			del indict["id"]
		except KeyError:
			raise TableError("You must specify an id to edit")
		values = []
		query = "UPDATE " + self.table + " SET "
		for key in indict:
			query = query + key + "=?, "
			values.append(indict[key])
		query = query[:-2] + " WHERE id=?;"
		values.append(id)
		self.sqlite(query, values)
		return self.print({'table':self.table, 'id':id})

	def search(self, indict):
		"""Search for text"""
		self.deftable(indict)
		#query = "SELECT id, "
		query = "SELECT "
		for column in self.tables[self.table]:
			query = query + column + ", "
		query = query[:-2] + " FROM " + self.table + " where "
		values = []
		for key in indict:
			query = query + key + " like ? AND "
			values.append("%" + indict[key] + "%")
		query = query[:-4] + ";"	

		self.sqlite(query, values)
		return self.makedict()

	def head(self, indict):
		"""Search head id from bookid and page"""
		query = "SELECT * FROM heads WHERE bid = ? AND hp <= ? ORDER BY hp;"
		self.sqlite(query, (indict["bid"], indict["p"]))
		self.table='heads'
		self.result=(self.result[-1],) # last
		return self.makedict()

	def print(self, indict):
		"""Print content of row or whole table"""
		self.deftable(indict)
		try:
			# if an id is defined, print row
			query = "SELECT * FROM " + self.table + " WHERE id = ?"
			self.sqlite(query, (str(indict["id"]),))
		except KeyError:
			# else, print all table
			query = "SELECT * FROM " + self.table
			self.sqlite(query)
		return self.makedict()



def main(db):
	sophia(db)

if __name__ == '__main__':
	status = main()
	sys.exit(status)
